File: laba3/task1/Golang/twolinkedlist.go
package main
import "fmt"

type TLList struct {
	head *Node
	tail *Node
}

func (this *TLList) createNodeTL(data string) *Node {
	return createNode(data)
}

func (this *TLList) addBackTL(data string) {
	node := this.createNodeTL(data)
	if this.head == nil {
		this.head = node
		this.tail = node
		return
	}
	node.prev = this.tail
	this.tail.next = node
	this.tail = node
}

func (this *TLList) remFrontTL() {
	if this.head == nil { return }
	this.head = this.head.next
	if this.head == nil {
		this.tail = nil
	} else {
		this.head.prev = nil
	}
}

func (this *TLList) remBackTL() {
	if this.head == nil { return }
	if this.head == this.tail {
		this.remFrontTL()
		return
	}
	this.tail = this.tail.prev
	this.tail.next = nil
}

func (this *TLList) printListTL() {
	if this.head == nil { return }
	curr := this.head
	for curr != nil {
		fmt.Print(curr.data, " ")
		curr = curr.next
	}
	fmt.Println()
}

func (this *TLList) addFrontTL(data string) {
	node := this.createNodeTL(data)
	if this.head == nil {
		this.head = node
		this.tail = node
		return
	}
	node.next = this.head
	this.head.prev = node
	this.head = node
}

func (this *TLList) insertAfterIndTL(index int, data string) {
	if index < 0 || this.head == nil { return }
	curr := this.head
	i := 0
	for ; i != index && curr != nil; i++ {
		curr = curr.next
	}
	if curr == nil { return }
	node := this.createNodeTL(data)
	node.next = curr.next
	node.prev = curr
	if curr.next != nil {
		curr.next.prev = node
	}
	curr.next = node
	if curr == this.tail {
		this.tail = node
	}
}

func (this *TLList) insertBeforeIndTL(index int, data string) {
	if index == 0 {
		this.addFrontTL(data)
		return
	}
	if index < 0 || this.head == nil { return }
	curr := this.head
	i := 0
	for ; i != index && curr != nil; i++ {
		curr = curr.next
	}
	if curr == nil { return }
	node := this.createNodeTL(data)
	node.prev = curr.prev
	node.next = curr
	curr.prev.next = node
	curr.prev = node
}

func (this *TLList) remValTL(val string) {
	if this.head == nil { return }
	if this.head.data == val {
		this.remFrontTL()
		return
	}
	curr := this.head
	for curr != nil && curr.data != val {
		curr = curr.next
	}
	if curr == nil { return }
	curr.prev.next = curr.next
	if curr == this.tail {
		this.tail = curr.prev
	} else {
		curr.next.prev = curr.prev
	}
}

func (this *TLList) findValTL(val string) int {
	if this.head == nil { return -1 }
	curr := this.head
	ind := 0
	for curr != nil && curr.data != val {
		curr = curr.next
		ind++
	}
	if curr == nil { return -1 }
	return ind
}

func (this *TLList) deleteBeforeIndTL(index int) {
	if this.head == nil {
		fmt.Println("Список пуст.")
		return
	}
	if index == 1 {
		this.remFrontTL()
		return
	}
	if index <= 0 {
		fmt.Println("Неверный индекс.")
		return
	}
	curr := this.head
	i := 0
	for ; i != index && curr != nil; i++ {
		curr = curr.next
	}
	if curr == nil || curr.prev == nil {
		fmt.Println("Неверный индекс.")
		return
	}
	node := curr.prev
	curr.prev = node.prev
	if node.prev != nil {
		node.prev.next = curr
	} else {
		this.head = curr
	}
}

func (this *TLList) deleteAfterIndTL(index int) {
	if this.head == nil || this.head == this.tail {
		fmt.Println("Удаление невозможно.")
		return
	}
	if index < 0 {
		fmt.Println("Неверный индекс.")
		return
	}
	curr := this.head
	i := 0
	for ; i != index && curr != nil; i++ {
		curr = curr.next
	}
	if curr == nil || curr.next == nil {
		fmt.Println("Неверный индекс.")
		return
	}
	node := curr.next
	curr.next = node.next
	if node.next != nil {
		node.next.prev = curr
	} else {
		this.tail = curr
	}
}

func (this *TLList) getHead() *Node { return this.head }
func (this *TLList) getTail() *Node { return this.tail }

File: laba3/task2/twolinkedlist.cpp
#include <iostream>
#include "structures.h"

using namespace std;

void TLList::addBackTL(const string& data)
{
    Node* node = createNodeTL(data);
    if (this->head == nullptr){
    	this->head = node;
    	this->tail = node;
    	return;
    }
    node->prev = this->tail; 
    this->tail->next = node;
    this->tail = node;
}

void TLList::remFrontTL()
{
	if (this->head == nullptr) return;
	const Node* temp = this->head;
	this->head = this->head->next;
	
	if (this->head == nullptr){
		this->tail = nullptr;
	} else {
		this->head->prev = nullptr;
	}
	delete temp;
}

void TLList::remBackTL()
{
    if (this->head == nullptr) return;
    if (this->head == this->tail) {remFrontTL(); return;}
    const Node* temp = this->tail;
    this->tail = this->tail->prev;
    this->tail->next = nullptr;
    delete temp;
}

void TLList::printListTL() const
{
	if (this->head == nullptr) return;
    const Node* curr = this->head;
    while (curr != nullptr) {
        cout << curr->data << " ";
        curr = curr->next;
    }
    cout << endl;
}


void TLList::addFrontTL(const string& data){
	Node* node = createNodeTL(data);
	if (this->head == nullptr){
    	this->head = node;
    	this->tail = node;
    	return;
    }
	node->next = this->head;
	this->head->prev = node;
	this->head = node;
}

void TLList::insertAfterIndTL(const int index, const string& data){
	if (index < 0|| this->head == nullptr) return;
	Node* curr = this->head;
	int i = 0;
	for (; i != index && curr != nullptr; i++){
		curr = curr->next;
	}
	if (curr == nullptr) return;
	Node* node = createNodeTL(data);
	node->next = curr->next;
	node->prev = curr;
	if (curr->next != nullptr) { curr->next->prev = node;}
	curr->next = node;
	if (curr == this->tail){
		this->tail = node;
	}
}


void TLList::insertBeforeIndTL(const int index, const string& data){
	if (index == 0) {addFrontTL(data); return;}
	if (index < 0|| this->head == nullptr) return;
	Node* curr = this->head;
	int i = 0;
	for (; i != index && curr != nullptr; i++){
		curr = curr->next;
	}
	if (curr == nullptr) return;
	Node* node = createNodeTL(data);
	node->prev = curr->prev;
	node->next = curr;
	curr->prev->next = node;
	curr->prev = node;
}

void TLList::remValTL(const string& val){
	if (this->head == nullptr) return;
    if (this->head->data == val) {remFrontTL(); return;}
    const Node* curr = this->head;
    while (curr != nullptr && curr->data != val){
    	curr = curr->next;
    }
    if (curr == nullptr) return;
    curr->prev->next = curr->next;
    if (curr == this->tail){
		this->tail = curr->prev;
	} else {
		curr->next->prev = curr->prev;
	}
	delete curr;
}

int TLList::findValTL(const string& val) const
{
	if (this->head == nullptr) return -1;
	const Node* curr = this->head;
	int ind = 0;
    while (curr != nullptr && curr->data != val){
    	curr = curr->next;
    	ind++;
    }
    if (curr == nullptr) return -1;
    return ind;
}

void TLList::deleteBeforeIndTL(const int index) {
    if (this->head == nullptr) {
        cout << "Список пуст.\n";
        return;
    }
    if (index == 1) {
	    remFrontTL();
	    return;
    }
    if (index <= 0) {
	    cout << "Неверный индекс.\n";
	    return;
    }
    Node* curr = this->head;
    int i = 0;
    for (; i != index && curr != nullptr; i++) {
        curr = curr->next;
    }
    if (curr == nullptr || curr->prev == nullptr) {
        cout << "Неверный индекс.\n";
        return;
    }
    const Node* node = curr->prev;
    curr->prev = node->prev;
    if (node->prev != nullptr){
    	node->prev->next = curr;
    } else {
    	this->head = curr;
    }
    delete node;
}

void TLList::deleteAfterIndTL(const int index) {
    if (this->head == nullptr || this->head == this->tail) {
        cout << "Удаление невозможно.\n";
        return;
    }
    if (index < 0) {
        cout << "Неверный индекс.\n";
        return;
    }
    Node* curr = this->head;
    int i = 0;
    for (; i != index && curr != nullptr; i++) {
        curr = curr->next;
    }
    if (curr == nullptr || curr->next == nullptr) {
        cout << "Неверный индекс.\n";
        return;
    }
    const Node* node = curr -> next;
    curr -> next = node -> next;
    if (node->next != nullptr) {
        node->next->prev = curr;
    } else {
    	this->tail = curr;
    }
    delete node;
}
